### ISSUE TITLE
Rootless Container Port Publishing via 127.0.0.1 Issue

### ISSUE THREAD

**Issue by bcstinch** [containers/libpod#5740]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

When trying to publish a port on a rootless container via 127.0.0.1, that port binding crashes when other containers try and access that port.

For example, I am using a postgres container to publish 127.0.0.1:5432, and then literally trying to curl 127.0.0.1:5432, when this happens, podman port -l no longer lists the port binding. This only happens when trying to curl this port from another container. If I try and curl 127.0.0.1:5432 via the host OS, it works fine.

When the podman port -l no longer lists that port, the container is still running, not in a crashed state. It can even still access the internet, its just no longer publishing its port.

https://paste.centos.org/view/5e3e7f59

**Steps to reproduce the issue:**

1. `/usr/bin/podman run -dt -p 127.0.0.1:5432:5432 -e POSTGRES_PASSWORD=password -e POSTGRES_USER=mattermost -e POSTGRES_DB=mattermost   docker.io/postgres:12`

2. `podman port -l`
Notice the port binding exists

3. `curl 127.0.0.1:5432`
Notice we can curl that port from the host.

4. `podman run -it docker.io/mattermost/mattermost-team-edition curl 127.0.0.1:5432`
Notice we got an error as if that port doesn't exist

5. `podman port -l`
Notice that port mapping no longer exists.

**Describe the results you received:**
`podman port -l` no longer lists the port mapping and I cannot access that port from another container.

**Describe the results you expected:**
Containers should be able to access the host's localhost for rootless container communication.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
Version:            1.8.2
RemoteAPI Version:  1
Go Version:         go1.13.6
OS/Arch:            linux/amd64
```

**Output of `podman info --debug`:**

```
debug:
  compiler: gc
  git commit: ""
  go version: go1.13.6
  podman version: 1.8.2
host:
  BuildahVersion: 1.14.3
  CgroupVersion: v2
  Conmon:
    package: conmon-2.0.14-1.fc31.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.14, commit: 083a0be12178013d44ff51ceda3090ea741b6516'
  Distribution:
    distribution: fedora
    version: "31"
  IDMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  MemFree: 30520635392
  MemTotal: 33592004608
  OCIRuntime:
    name: crun
    package: crun-0.13-1.fc31.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 0.13
      commit: e79e4de4ac16da0ce48777afb72c6241de870525
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  SwapFree: 16865292288
  SwapTotal: 16865292288
  arch: amd64
  cpus: 8
  eventlogger: journald
  hostname: localhost.localdomain
  kernel: 5.3.7-301.fc31.x86_64
  os: linux
  rootless: true
  slirp4netns:
    Executable: /usr/bin/slirp4netns
    Package: slirp4netns-0.4.0-20.1.dev.gitbbd6f25.fc31.x86_64
    Version: |-
      slirp4netns version 0.4.0-beta.3+dev
      commit: bbd6f25c70d5db2a1cd3bfb0416a8db99a75ed7e
  uptime: 10h 6m 3.68s (Approximately 0.42 days)
registries:
  search:
  - docker.io
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - quay.io
store:
  ConfigFile: /home/bcs/.config/containers/storage.conf
  ContainerStore:
    number: 4
  GraphDriverName: overlay
  GraphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-0.7.8-1.fc31.x86_64
      Version: |-
        fusermount3 version: 3.6.2
        fuse-overlayfs: version 0.7.8
        FUSE library version 3.6.2
        using FUSE kernel interface version 7.29
  GraphRoot: /home/bcs/.local/share/containers/storage
  GraphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  ImageStore:
    number: 2
  RunRoot: /run/user/1000/containers
  VolumePath: /home/bcs/.local/share/containers/storage/volumes
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-1.8.2-2.fc31.x86_64
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Bare metal with Fedora Server 31 installed.

**Comment by storrgie** [comment]
All the subsystems are new to me, so I could be guessing completely wrong. Does this mean that slirp4netns is "failing silently" and it is responsible for that forwarding of the port?

**Comment by mheon** [comment]
Per debugging yesterday, I suspect that slirp is still up, but the rootlesskit port forwarder is dying.

@giuseppe @AkihiroSuda Mind taking a look at this one?

**Comment by AkihiroSuda** [comment]
Please provide the full reproducer. Both command 1 and 3 don't seem correct.

**Comment by bcstinch** [comment]
> 
> 
> Please provide the full reproducer. Both command 1 and 3 don't seem correct.

I apologize, I forgot I use an env-file for the postgres stuff. I have edited the original post.

**Comment by AkihiroSuda** [comment]
> When trying to publish a port on a rootless container via 127.0.0.1, that port binding crashes when other containers try and access that port.

The port binding seems crashing just when yet another container started up.
So command 4 can be just like `podman run hello-world`. And anyway you can't connect 127.0.0.1 from  a container (regardless to rootless or rootful) unless you specify `--net=host`.

**Comment by AkihiroSuda** [comment]
The port itself seems alive, but `podman port -l` CLI seems broken.

**Comment by bcstinch** [comment]
With that in mind, what is the preferred method for communicating between two containers via localhost so that traffic doesn't have to "leave the host" or open up any ports on a network-facing IP?

We were originally putting them in the same pod, but we were seeing that if we restart a container within that pod, it cannot re-bind to a published port unless we restart the _entire_ pod, which is not ideal.

**Comment by AkihiroSuda** [comment]
I suggest using UNIX sockets

**Comment by storrgie** [comment]
> I suggest using UNIX sockets

Could you elaborate further? Some of the things you'd mentioned are not extremely obvious (e.g. the `--net=host` thing I'd not even spotted until you'd mentioned it).

**Comment by AkihiroSuda** [comment]
See https://docs.mattermost.com/install/sockets-db.html . You will need to create shared volume for `/run`.

> the --net=host thing 

In Rootless Podman, each of containers has its own network namespace. i.e. each of containers have  its own "localhost" which is different from "the real localhost". 
`podman run --net=host` disables network namespace so that the container can access "the real localhost" as "localhost".

**Comment by storrgie** [comment]
I think this can be closed as we've posted #5828.
